diff -ur ../../ns-allinone-2.34.orig/tclcl-1.19/conf/configure.in.tcl ./conf/configure.in.tcl
--- ../../ns-allinone-2.34.orig/tclcl-1.19/conf/configure.in.tcl	2009-06-14 11:35:03.000000000 -0600
+++ ./conf/configure.in.tcl	2012-03-30 10:58:10.000000000 -0600
@@ -99,6 +99,9 @@
 		$d/include/tcl$TCL_HI_VERS \
 		$d/include/tcl$TCL_VERS \
 		$d/include/tcl$TCL_ALT_VERS \
+                $d/include/tcl$TCL_HI_VERS/generic \
+                $d/include/tcl$TCL_VERS/generic \
+                $d/include/tcl$TCL_ALT_VERS/generic \
 		$d/include \
 		/usr/local/include \ 
 		"
diff -ur ../../ns-allinone-2.34.orig/tclcl-1.19/configure ./configure
--- ../../ns-allinone-2.34.orig/tclcl-1.19/configure	2009-06-14 11:35:03.000000000 -0600
+++ ./configure	2012-03-30 10:58:10.000000000 -0600
@@ -6014,6 +6014,9 @@
 		$d/include/tcl$TCL_HI_VERS \
 		$d/include/tcl$TCL_VERS \
 		$d/include/tcl$TCL_ALT_VERS \
+		$d/include/tcl$TCL_HI_VERS/generic \
+		$d/include/tcl$TCL_VERS/generic \
+		$d/include/tcl$TCL_ALT_VERS/generic \
 		$d/include \
 		/usr/local/include \
 		"
