--- libmisc/chkname.c~	2017-11-21 15:59:55.560838330 -0700
+++ libmisc/chkname.c	2017-11-21 16:01:17.520596721 -0700
@@ -52,12 +52,13 @@
 	 * User/group names must match [a-z_][a-z0-9_-]*[$]
 	 */
 	if (('\0' == *name) ||
-	    !((('a' <= *name) && ('z' >= *name)) || ('_' == *name))) {
+	    !((('a' <= *name) && ('z' >= *name)) || (('A' <= *name) && ('Z' >= *name)) || ('_' == *name))) {
 		return false;
 	}
 
 	while ('\0' != *++name) {
 		if (!(( ('a' <= *name) && ('z' >= *name) ) ||
+		      ( ('A' <= *name) && ('Z' >= *name) ) ||
 		      ( ('0' <= *name) && ('9' >= *name) ) ||
 		      ('_' == *name) ||
 		      ('-' == *name) ||
