*** networking.sh     2018-04-04 06:23:20.910078008 -0600
--- /etc/init.d/networking.sh       2018-04-04 10:13:19.472273061 -0600
***************
*** 196,201 ****
--- 196,208 ----
      intf_list=$(net_intf)
      intf_counter=0

+     # Added by Emulab
+     PLATFORM=`onie-syseeprom -g 0x2d`
+     rc=$?
+     if [ $rc -eq 0 -a "$PLATFORM" = "DELL" ]; then
+       intf_counter=2
+     fi
+ 
      # Set MAC addr for all interfaces, but leave the interfaces down.
      base_mac=$(onie-sysinfo -e)
      for intf in $intf_list ; do
