--- Makefile.old	2012-05-22 13:44:22.000000000 -0600
+++ Makefile	2012-05-22 13:44:30.000000000 -0600
@@ -563,5 +563,30 @@
     SUBDIR += ytree
     SUBDIR += zaptel
     SUBDIR += zoneinfo
+    SUBDIR += emulab-boss
+    SUBDIR += emulab-clrhouse
+    SUBDIR += emulab-dhcp31-client
+    SUBDIR += emulab-extras
+    SUBDIR += emulab-fs
+    SUBDIR += emulab-iperf
+    SUBDIR += emulab-mailman
+    SUBDIR += emulab-ops
+    SUBDIR += emulab-php4
+    SUBDIR += emulab-php5
+    SUBDIR += emulab-protogeni
+    SUBDIR += emulab-pubsub
+    SUBDIR += emulab-py-m2crypto
+    SUBDIR += emulab-tftp-hpa
+    SUBDIR += emulab-tipserv
+    SUBDIR += libtelnet
+    SUBDIR += net-snmp4
+    SUBDIR += p5-DBD-mysql50-old
+    SUBDIR += p5-GD-old
+    SUBDIR += p5-IO-EventMux
+    SUBDIR += p5-IO-EventMux-Socket-MsgHdr
+    SUBDIR += p5-Privileges-Drop
+    SUBDIR += p5-RPC-Async
+    SUBDIR += p5-SNMP
+    SUBDIR += ulsshxmlrpcpp
 
 .include <bsd.port.subdir.mk>
